tests/integration/telemetry/stackdriver/vm/main_test.go

```diff
@@ -113,7 +189,156 @@ func testSetup(ctx resource.Context) (err error) {
 		"ISTIO_METAJSON_LABELS":                                  vmLabelsJSON,
 		"GCE_METADATA_HOST":                                      gceInst.Address(),
+		"CANONICAL_SERVICE":                                      "vm-server",
+		"CANONICAL_REVISION":                                     "v1",
+		// we must supply a bootstrap override to get the test endpoint uri into the tracing configuration
+		"ISTIO_BOOTSTRAP_OVERRIDE": "/etc/istio/custom-bootstrap/custom_bootstrap.json",
+	}
+
+	// read expected values from testdata
+	wantClientReqs, wantServerReqs, err = goldenRequestCounts()
+	if err != nil {
+		return fmt.Errorf("failed to get golden metrics from file: %v", err)
+	}
+	wantLogEntry, err = goldenLogEntry()
+	if err != nil {
+		return fmt.Errorf("failed to get golden log entry from file: %v", err)
+	}
+	wantTrafficAssertion, err = goldenTrafficAssertion()
+	if err != nil {
+		return fmt.Errorf("failed to get golden traffic assertion from file: %v", err)
+	}
+	wantTrace, err = goldenTrace()
+	if err != nil {
+		return fmt.Errorf("failed to get golden trace from file: %v", err)
+	}
+
+	// set up client and server
+	ports := []echo.Port{
+		{
+			Name:     "http",
+			Protocol: protocol.HTTP,
+			// Due to a bug in WorkloadEntry, service port must equal target port for now
+			InstancePort: 8090,
+			ServicePort:  8090,
+		},
+	}
+
+	// builder to build the instances iteratively
+	clientBuilder = echoboot.NewBuilder(ctx).
+		With(&client, echo.Config{
+			Service:   "client",
+			Namespace: ns,
+			Ports:     ports,
+			Subsets: []echo.SubsetConfig{
+				{
+					Annotations: echo.NewAnnotations().Set(echo.SidecarBootstrapOverride, sdBootstrapConfigMap),
+				},
+			},
+		})
+
+	serverBuilder = echoboot.NewBuilder(ctx).
+		With(&server, echo.Config{
+			Service:       "server",
+			Namespace:     ns,
+			Ports:         ports,
+			DeployAsVM:    true,
+			VMEnvironment: vmEnv,
+			Subsets: []echo.SubsetConfig{
+				{
+					Annotations: echo.NewAnnotations().
+						Set(proxyConfigAnnotation, envTagsProxyConfig).
+						Set(echo.SidecarBootstrapOverride, sdBootstrapConfigMap),
+				},
+			},
+		})
+
+	return nil
+}
+
+func goldenRequestCounts() (cltRequestCount, srvRequestCount *monitoring.TimeSeries, err error) {
+	srvRequestCountTmpl, err := ioutil.ReadFile(serverRequestCount)
+	if err != nil {
+		return
 	}
+	sr, err := tmpl.Evaluate(string(srvRequestCountTmpl), map[string]interface{}{
+		"EchoNamespace": ns.Name(),
+	})
+	if err != nil {
+		return
+	}
+	cltRequestCount = &monitoring.TimeSeries{}
+	srvRequestCount = &monitoring.TimeSeries{}
+	if err = jsonpb.UnmarshalString(sr, srvRequestCount); err != nil {
+		return
+	}
+	cltRequestCountTmpl, err := ioutil.ReadFile(clientRequestCount)
+	if err != nil {
+		return
+	}
+	cr, err := tmpl.Evaluate(string(cltRequestCountTmpl), map[string]interface{}{
+		"EchoNamespace": ns.Name(),
+	})
+	if err != nil {
+		return
+	}
+	err = jsonpb.UnmarshalString(cr, cltRequestCount)
```

Comment:
nit: move declaration of cltRequestCount here.. so it's closer to where it's getting used?